xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingExchange.java

```diff
@@ -18,47 +17,72 @@
 public class KrakenStreamingExchange extends KrakenExchange implements StreamingExchange {
 
     private static final Logger LOG = LoggerFactory.getLogger(KrakenStreamingExchange.class);
-
     private static final String API_URI = "wss://ws.kraken.com";
+    private static final String API_AUTH_URI = "wss://ws-auth.kraken.com";
+    private static final String API_BETA_URI = "wss://beta-ws.kraken.com";
 
-    private final KrakenStreamingService streamingService;
+    private KrakenStreamingService streamingService, privateStreamingService;
     private KrakenStreamingMarketDataService streamingMarketDataService;
-    
+    private KrakenStreamingTradeService streamingTradeService;
+
     private final SynchronizedValueFactory<Long> nonceFactory = new CurrentTimeNonceFactory();
 
     public KrakenStreamingExchange() {
-        this.streamingService = new KrakenStreamingService(API_URI);
+    }
+
+    private static String pickUri(boolean isPrivate, Boolean useSandbox) {
+        if (useSandbox != null && useSandbox)
+            return API_BETA_URI;
+        else
+            return isPrivate ? API_AUTH_URI : API_URI;
     }
 
     @Override
     protected void initServices() {
         super.initServices();
-        streamingMarketDataService = new KrakenStreamingMarketDataService(streamingService);
+        Boolean useSanbox = (Boolean)exchangeSpecification.getExchangeSpecificParametersItem(USE_SANDBOX);
+
+        this.streamingService = new KrakenStreamingService(false, pickUri(false,useSanbox));
+        this.streamingMarketDataService = new KrakenStreamingMarketDataService(streamingService);
+
+        if (exchangeSpecification.getApiKey() != null) {
```

Comment:
Use `StringUtils.isNotEmpty`; when substituting environment variables into application config, it's hard to force a configuration value to null.